### PR TITLE
User course input

### PR DESCRIPTION
**Pull request by aldy97**
# Changes
- User can now add valid courses to Redux
- Textfield can auto-correct lower case courseDepartment input to uppercase ('cpsc 121' => 'CPSC 121', which is how the course is stored in the db)
- There will be a toast message rendered at top center when user attempts to add a course, message content is varied based on different scenarios.

## Motivation and Context
- We need to store selected courses to prepare for generaiting course schedule later on
- We have to make sure that all courses recorded do exist in the db

## How Has This Been Tested?
- for example, user's input 'cpsc 221' would be corrected and added as 'CPSC 221'
- if the course does not exist in db, it wont be added
- if a validated course has been added, it wont be added again.
- When navigating back and forth between different stages, it is able to remember courses stored in redux and render them properly.

## Screenshots (if appropriate):
image
image
image

# Related Issues
- This is a back-up solution for autocomplete
- When too many courses have been added (>8), it does not render properly.
- Added courses are not yet be able to be removed for now.

### CODE REVIEW

File: client/src/components/Courses.tsx [hunk 1]
```diff
@@ -26,23 +34,124 @@ const CourseList = styled.div`
...
+    if (partition.length !== 2) {
+      setMessage("Invalid format");
+      setOpen(true);
+      return;
+    }
+
+    const department = partition[0].toUpperCase();
+    const courseNumber = partition[1];
+    const response = await axios.get(
+      `http://localhost:5000/api/section/${department}/${courseNumber}`
```
Comment (johnagl) [inline client/src/components/Courses.tsx:73]: Once my PR is merged we can query `http://localhost:5000/api/section/${term}/${department}/${courseNumber}`

**Comment by johnagl** [comment]
coursesAdded

Instead of just coursedept + coursenumber string, we should add all course sections queried from db, so it should basically be an array of objects. We don't want to query the database again later. We can do this once we have the endpoint with term merged.

For example, 1 courseSection is : 
    {
    coursetitle: "CPSC 100 Computational Thinking",
    coursedept: "CPSC",
    coursenumber: "100",
    sectiontitle: "CPSC 100 L1C",
    status: "",
    activity: "Laboratory",
    prof: "",
    term: "1",
    day: "Tue",
    starttime: "17:30",
    endtime: "18:30"
    }

Other than that, everything looks good!

**Comment by aldy97** [comment]
> coursesAdded
> 
> Instead of just coursedept + coursenumber string, we should add all course sections queried from db, so it should basically be an array of objects. We don't want to query the database again later. We can do this once we have the endpoint with term merged.
> 
> For example, 1 courseSection is :
> {
> coursetitle: "CPSC 100 Computational Thinking",
> coursedept: "CPSC",
> coursenumber: "100",
> sectiontitle: "CPSC 100 L1C",
> status: "",
> activity: "Laboratory",
> prof: "",
> term: "1",
> day: "Tue",
> starttime: "17:30",
> endtime: "18:30"
> }
> 
> Other than that, everything looks good!

Okay, sounds great. I have a question: so for each course added, we will have an array of objects and in the array we will inlcude lectures, tuorials and labs for that one specific course?

**Comment by johnagl** [comment]
> Okay, sounds great. I have a question: so for each course added, we will have an array of objects and in the array we will inlcude lectures, tuorials and labs for that one specific course?

I think we can have a state in reduxStore which is an array of CourseSection objects. So [CourseSection 1, CourseSection2, CourseSection3.... CourseSectionX] etc. The reason for this is, you're querying the database, so might as well grab the course section information so we don't have to do it again later. We use this array of course sections to generate course schedules later.  

You can put all course sections into 1 array in redux store.

**Comment by johnagl** [comment]
And yes it includes lectures, labs , tutorials, everything XD. (At least for now)

**Comment by aldy97** [comment]
> And yes it includes lectures, labs , tutorials, everything XD. (At least for now)

**Pull request closed without merging**